### DISCUSSION TITLE
about intrumentResample and instrumentArcIntervals

### DISCUSSION THREAD

**Discussion by zdwangjian5** [groops-devs/groops#267] [Q&A] [answered]
Please help me solve this problem. I want to interpolate CLK1B to every moment of KBR1A/LRI1A, and found that when directly using instrumentResample to linearly interpolate CLK1B to KBR1A directly, there will be some NAN values. How should I interpolate it correctly? Does it have something to do with the fact that CLK1B sometimes has jumps, and whether it should be piecewise interpolation. I saw that there is a way of instrumentArcIntervals in instrumentResample, is this piecewise interpolation for each arc segment? However, after I tested it, I found that I could not get the desired result. Please help me, thank you.

**Comment by felingeroehlix** [comment]
Hi zdwangjian5,

Have you converted both files correctly into GROOPS instrument file format?. Besides that, there should be no problem to interpolate CLK1B data to any other data using InstrumentResample (timeSeries: instrument). I hope this answers your question.

Best regards,
Felix

**Reply by zdwangjian5** [reply]
Yes,I have converted both files correctly into GROOPS instrument file format,but there will be some NAN values.

**Reply by zdwangjian5** [reply]
Maybe it is becaused i set the wrong extrapolation and range parameters due to time jumps? The extrapolation in matlab is a boolean value, but here i have to set a value. I am not very clear about how to set it through the discription

**Reply by felingeroehlix** [reply]
Can you send me your example scenario? I cannot completely follow what your problem is. It  will probably be the settings of InstrumentResample. (Settings of [InterpolatorTimeSeries](https://groops-devs.github.io/groops/html/interpolatorTimeSeriesType.html))

**Reply by zdwangjian5** [reply]
The LRI1A file is too large after zip. I can not upload it here. The example datas are just the clk1b 2019-11-01_c.1b.txt and LRI 2019-11-01_C.txt from JPL ftp. I have converted both files correctly into GROOPS instrument file format,but there will be some NAN values. I have another more urgent question about interpolation to consult you: My usage scenario is like this. I assume a time series where there are outliers at some consecutive moments in the middle. I need to use the data before and after the outliers to interpolate the abnormal moments and replace the original outliers with the interpolated values. I originally wanted to use polynomial interpolation. The code is as follows: Polynomial polynomial(timesOld, degree); const Matrix fc = polynomial.interpolate(timesNew, interpData1); When directly calling Polynomial, it is also impossible to set range and extrapolation. However, the interpolate function makes an error when judging dt2 and exception, and enters std::exception. How should I correctly call the interpolation function in the code? Thank you very much!

**Reply by felingeroehlix** [reply, marked as answer]
For your first question. Using GraceL1b2ClockOffset for the CLK1B data and GraceL1a2SatelliteTracking for the KBR1A it is no problem to resample the CLK1B data to the time series of KBR1A using InstrumentResample (method: polynomial degree 1, timeSeries: instrument). However,  GraceL1a2SatelliteTracking is only designed for KBR data and you have to adjust the code accordingly for LRI data. This is probably where your NAN values are coming from. 

For your second question. You need to remove your outliers before you interpolate the time series. You can use e.g. InstrumentRemoveEpochsByCriteria and set a certain threshold condition (or however you define an outlier) and then use InstrumentResample (method: fillGapsLeastSquaresPolynomialFit).

**Reply by zdwangjian5** [reply]
Thank you for your help!
